File: tests/oracle/rule_earlyremoval_1204.cc
/*
 * RULE-PORT-EARLYREMOVAL-0001
 *
 * Locked Ghidra 12.0.4 oracle for RuleEarlyRemoval::applyOp,
 * Heritage::deadRemovalAllowedSeen, AddrSpace::doesDeadcode, and the default
 * Rule::getOpList dispatch contract.  Every mutation is performed through
 * Funcdata/PcodeOp/Varnode public APIs.  The runner adds three fixture-only
 * Heritage state accessors to its immutable archive so pass and deadremoved
 * can be observed without changing any production algorithm.
 */

#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "ruleaction.hh"

#include <algorithm>
#include <iostream>
#include <iterator>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::ostringstream;
using std::string;
using std::vector;

static bool containsOp(list<PcodeOp *>::const_iterator iter,
                       list<PcodeOp *>::const_iterator enditer,
                       const PcodeOp *needle)
{
  for(;iter!=enditer;++iter)
    if (*iter == needle) return true;
  return false;
}

static bool containsVarnode(const Funcdata &fd,const Varnode *needle)
{
  if (needle == (const Varnode *)0) return false;
  for(VarnodeLocSet::const_iterator iter=fd.beginLoc();iter!=fd.endLoc();++iter)
    if (*iter == needle) return true;
  return false;
}

static bool containsBlockOp(const BlockBasic *block,const PcodeOp *needle)
{
  for(list<PcodeOp *>::const_iterator iter=block->beginOp();
      iter!=block->endOp();++iter)
    if (*iter == needle) return true;
  return false;
}

class Graph {
public:
  Funcdata fd;
  Architecture &arch;
  AddrSpace *code;
  AddrSpace *unique;
  AddrSpace *other;
  BlockBasic *block;
  uintb nextPc;
  uintb nextOther;

  Graph(Architecture &a,const string &name,uintb base,int4 pass)
    : fd(name,name,a.symboltab->getGlobalScope(),
         Address(a.getDefaultCodeSpace(),base),(FunctionSymbol *)0,0x100),
      arch(a),code(a.getDefaultCodeSpace()),unique(a.getUniqueSpace()),
      other(a.getSpaceByName("OTHER")),block((BlockBasic *)0),
      nextPc(base),nextOther(0x9000)
  {
    if (other == (AddrSpace *)0)
      other = a.getSpaceByName("other");
    if (code == (AddrSpace *)0 || unique == (AddrSpace *)0 ||
        other == (AddrSpace *)0)
      throw std::runtime_error("required code/unique/OTHER space is absent");
    BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
    block = blocks.newBlockBasic(&fd);
    fd.fixtureSetHeritagePass(pass);
  }

  PcodeOp *newOp(OpCode opcode,int4 inputCount,bool withOutput,AddrSpace *space)
  {
    PcodeOp *op = fd.newOp(inputCount,Address(code,nextPc++));
    fd.opSetOpcode(op,opcode);
    if (inputCount > 0)
      fd.opSetInput(op,fd.newConstant(4,nextPc),0);
    if (withOutput) {
      if (space == unique)
        fd.newUniqueOut(4,op);
      else {
        Varnode *out = fd.newVarnode(4,Address(space,nextOther++));
        fd.opSetOutput(op,out);
      }
    }
    fd.opInsertEnd(op,block);
    return op;
  }
};

struct CaseSpec {
  const char *name;
  OpCode opcode;
  int4 pass;
  int4 deadDelay;
  bool withOutput;
  bool indirectSource;
  bool descendant;
  bool writeMask;
  bool autoLive;
  bool otherSpace;
};

static void runCase(Architecture &arch,const CaseSpec &spec,uintb base)
{
  Graph graph(arch,spec.name,base,spec.pass);
  AddrSpace *outSpace = spec.otherSpace ? graph.other : graph.unique;
  graph.fd.setDeadCodeDelay(outSpace,spec.deadDelay);
  PcodeOp *op = graph.newOp(spec.opcode,1,spec.withOutput,outSpace);
  if (spec.indirectSource)
    op->setIndirectSource();
  Varnode *input = op->getIn(0);
  Varnode *output = op->getOut();
  if (output != (Varnode *)0) {
    if (spec.writeMask) output->setWriteMask();
    if (spec.autoLive) output->setAutoLiveHold();
  }
  if (spec.descendant) {
    if (output == (Varnode *)0)
      throw std::runtime_error("descendant case requires an output");
    PcodeOp *sink = graph.newOp(CPUI_COPY,1,true,graph.unique);
    graph.fd.opSetInput(sink,output,0);
  }

  const int4 descBefore = output == (Varnode *)0 ? -1 :
    (int4)std::distance(output->beginDescend(),output->endDescend());
  const int4 inputDescBefore =
    (int4)std::distance(input->beginDescend(),input->endDescend());
  const bool callBefore = op->isCall();
  const bool indirectBefore = op->isIndirectSource();
  const bool writeBefore = output != (Varnode *)0 && output->isWriteMask();
  const bool autoBefore = output != (Varnode *)0 && output->isAutoLive();
  const bool spaceDeadcode = output != (Varnode *)0 &&
    output->getSpace()->doesDeadcode();

  RuleEarlyRemoval rule("deadcode");
  int4 result = rule.applyOp(op,graph.fd);
  const int4 inputDescAfter =
    (int4)std::distance(input->beginDescend(),input->endDescend());
  const bool alive = containsOp(graph.fd.beginOpAlive(),graph.fd.endOpAlive(),op);
  const bool dead = containsOp(graph.fd.beginOpDead(),graph.fd.endOpDead(),op);
  const bool blockMember = containsBlockOp(graph.block,op);
  const bool outputPresent = containsVarnode(graph.fd,output);
  const bool outputAttached = op->getOut() != (Varnode *)0;
  const int4 deadremoved = graph.fd.fixtureGetDeadRemoved(outSpace);

  std::cout << "case=" << spec.name
            << "|result=" << result
            << "|pass=" << spec.pass
            << "|delay=" << spec.deadDelay
            << "|call=" << callBefore
            << "|indirect=" << indirectBefore
            << "|output=" << (output == (Varnode *)0 ? 0 : 1)
            << "|desc=" << descBefore
            << "|writemask=" << writeBefore
            << "|autolive=" << autoBefore
            << "|space_deadcode=" << spaceDeadcode
            << "|deadremoved=" << deadremoved
            << "|alive=" << alive
            << "|dead=" << dead
            << "|block=" << blockMember
            << "|output_present=" << outputPresent
            << "|output_attached=" << outputAttached
            << "|input_desc=" << inputDescBefore << "->" << inputDescAfter
            << '\n';

  if (string(spec.name) == "neither_allowed") {
    std::cout << "residual=nullable_input_arity"
              << "|post_num_inputs=" << op->numInput()
              << "|slot0_null=" << (op->getIn(0)==(Varnode *)0 ? 1 : 0)
              << "|status=MISMATCH" << '\n';
  }
}

static string joinOpcodes(const vector<uint4> &opcodes,bool typedOnly)
{
  ostringstream out;
  bool first = true;
  for(vector<uint4>::const_iterator iter=opcodes.begin();iter!=opcodes.end();++iter) {
    if (typedOnly && (*iter == 0 || *iter == 45)) continue;
    if (!first) out << ',';
    first = false;
    out << *iter;
  }
  return out.str();
}

static string aliveOpcodes(const Funcdata &fd)
{
  ostringstream out;
  bool first = true;
  for(list<PcodeOp *>::const_iterator iter=fd.beginOpAlive();
      iter!=fd.endOpAlive();++iter) {
    if (!first) out << ',';
    first = false;
    out << static_cast<int4>((*iter)->code());
  }
  return out.str();
}

static void runDispatch(Architecture &arch)
{
  Graph graph(arch,"typed_dispatch",0x7200,1);
  graph.fd.setDeadCodeDelay(graph.unique,0);
  RuleEarlyRemoval *rule = new RuleEarlyRemoval("deadcode");
  vector<uint4> raw;
  rule->getOpList(raw);
  vector<uint4> typed;
  for(vector<uint4>::const_iterator iter=raw.begin();iter!=raw.end();++iter)
    if (*iter != 0 && *iter != 45) typed.push_back(*iter);

  std::cout << "typed_oplist=" << joinOpcodes(raw,true)
            << "|count=" << typed.size() << '\n';
  std::cout << "residual=raw_dispatch|raw=" << joinOpcodes(raw,false)
            << "|count=" << raw.size()
            << "|untyped=0,45|status=MISMATCH" << '\n';

  for(vector<uint4>::const_iterator iter=typed.begin();iter!=typed.end();++iter)
    graph.newOp((OpCode)*iter,0,true,graph.unique);

  ActionPool pool(0,"earlyremoval_dispatch");
  pool.addRule(rule);
  int4 result = pool.perform(graph.fd);
  size_t deadCount = (size_t)std::distance(graph.fd.beginOpDead(),
                                           graph.fd.endOpDead());
  std::cout << "dispatch|perform=" << result
            << "|status=" << pool.getStatus()
            << "|tests=" << rule->getNumTests()
            << "|apply=" << rule->getNumApply()
            << "|alive=" << aliveOpcodes(graph.fd)
            << "|dead_count=" << deadCount
            << "|deadremoved="
            << graph.fd.fixtureGetDeadRemoved(graph.unique)
            << '\n';
}

static void run(Architecture &arch)
{
  const CaseSpec cases[] = {
    {"call_guard",CPUI_CALL,1,0,true,false,false,false,false,false},
    {"indirect_guard",CPUI_COPY,1,0,true,true,false,false,false,false},
    {"no_output_guard",CPUI_COPY,1,0,false,false,false,false,false,false},
    {"descendant_guard",CPUI_COPY,1,0,true,false,true,false,false,false},
    {"autolive_only",CPUI_COPY,1,0,true,false,false,false,true,false},
    {"write_and_autolive",CPUI_COPY,1,0,true,false,false,true,true,false},
    {"write_mask_only",CPUI_COPY,1,0,true,false,false,true,false,false},
    {"neither_allowed",CPUI_COPY,1,0,true,false,false,false,false,false},
    {"delay_zero_equal",CPUI_COPY,0,0,true,false,false,false,false,false},
    {"delay_one_equal",CPUI_COPY,1,1,true,false,false,false,false,false},
    {"delay_one_after",CPUI_COPY,2,1,true,false,false,false,false,false},
    {"no_deadcode_other",CPUI_COPY,0,0,true,false,false,false,false,true},
  };
  for(size_t i=0;i<sizeof(cases)/sizeof(cases[0]);++i)
    runCase(arch,cases[i],0x7000 + (uintb)i * 0x20);
  runDispatch(arch);
  std::cout << "residual=heritage_manager_projection"
            << "|scope=full_manager_state_and_late_generation_warning"
            << "|status=UNTESTED" << '\n';
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: rule_earlyremoval_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    vector<string> specPaths;
    specPaths.push_back(argv[1]);
    startDecompilerLibrary(specPaths);
    {
      std::ostringstream diagnostics;
      BfdArchitecture architecture(argv[2],"default",&diagnostics);
      DocumentStorage store;
      architecture.init(store);
      architecture.readLoaderSymbols("::");
      Funcdata *getstr = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
      if (getstr == (Funcdata *)0 || getstr->getAddress().getOffset() != 0x36d0)
        throw std::runtime_error("GetStr input identity drifted");
      if (architecture.archid != "x86:LE:64:default:gcc")
        throw std::runtime_error("runtime architecture/compiler drifted: " +
                                 architecture.archid);
      run(architecture);
    }
    shutdownDecompilerLibrary();
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/rule_earlyremoval_1204.rs
//! RULE-PORT-EARLYREMOVAL-0001: Rust twin of the locked Ghidra 12.0.4
//! RuleEarlyRemoval fixture.
//!
//! Covered records are compared byte-for-byte.  Records beginning with
//! `residual=` deliberately expose unrepresentable raw opcode buckets,
//! nullable post-destroy input slots, and the untested full Heritage-manager
//! closure; they are excluded from the narrow MATCH projection but retained
//! in the raw MISMATCH observation.

use std::sync::{Arc, RwLock};

use rugra::action::{Action, ActionPool, ActionState, Rule};
use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::block::FlowBlock;
use rugra::funcdata::Funcdata;
use rugra::op::{pcodeop_flags, PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::ruleaction::RuleEarlyRemoval;
use rugra::space::{AddressSpace, SPACEID_OTHER};
use rugra::varnode::Varnode;

type BlockRef = Arc<RwLock<dyn FlowBlock + Send + Sync>>;
type OpRef = Arc<RwLock<PcodeOp>>;
type VnRef = Arc<RwLock<Varnode>>;

struct Graph {
    fd: Funcdata,
    block: BlockRef,
    next_pc: u64,
    next_other: u64,
}

impl Graph {
    fn new(name: &str, base: u64, pass: i32) -> Self {
        let mut fd = Funcdata::new(name, Address::new(base), 0x100);
        let mut arch = Architecture::new();
        arch.archid = "x86:LE:64:default:gcc".to_string();
        fd.set_arch(Arc::new(arch));
        let block = fd.create_new_block();
        fd.heritage.build_info_list();
        fd.heritage.pass = pass;
        Self {
            fd,
            block,
            next_pc: base,
            next_other: 0x9000,
        }
    }

    fn new_op(
        &mut self,
        opcode: OpCode,
        input_count: usize,
        with_output: bool,
        space: AddressSpace,
    ) -> PcodeOpRef {
        let op = self.fd.new_op(input_count, Address::new(self.next_pc));
        self.next_pc += 1;
        self.fd.op_set_opcode(&op, opcode);
        if input_count > 0 {
            let input = self.fd.new_constant(4, self.next_pc);
            self.fd.op_set_input(&op, input, 0);
        }
        if with_output {
            if space == AddressSpace::Unique {
                self.fd.new_unique_out(4, &op);
            } else {
                let output = self.fd.vbank.create_with_space(4, space, self.next_other);
                self.next_other += 1;
                self.fd.op_set_output(&op, output);
            }
        }
        self.fd.op_insert_end(&op, &self.block);
        op
    }
}

#[derive(Clone, Copy)]
struct CaseSpec {
    name: &'static str,
    opcode: OpCode,
    pass: i32,
    dead_delay: i32,
    with_output: bool,
    indirect_source: bool,
    descendant: bool,
    write_mask: bool,
    auto_live: bool,
    other_space: bool,
}

fn bank_contains_op(bank: &[PcodeOpRef], needle: &OpRef) -> bool {
    bank.iter().any(|op| Arc::ptr_eq(&op.0, needle))
}

fn bank_contains_vn(fd: &Funcdata, needle: &VnRef) -> bool {
    fd.vbank
        .loc_tree
        .iter()
        .any(|vn| Arc::ptr_eq(&vn.0, needle))
}

fn block_contains(block: &BlockRef, needle: &OpRef) -> bool {
    block
        .read()
        .unwrap()
        .get_ops()
        .iter()
        .any(|op| Arc::ptr_eq(&op.0, needle))
}

fn dead_removed(fd: &Funcdata, space: AddressSpace) -> i32 {
    fd.heritage
        .infolist
        .iter()
        .find(|info| info.space == space)
        .map_or(-1, |info| info.deadremoved)
}

fn run_case(spec: CaseSpec, base: u64) {
    let mut graph = Graph::new(spec.name, base, spec.pass);
    let out_space = if spec.other_space {
        AddressSpace::Other(SPACEID_OTHER)
    } else {
        AddressSpace::Unique
    };
    graph
        .fd
        .heritage
        .set_dead_code_delay(out_space, spec.dead_delay);
    let op = graph.new_op(spec.opcode, 1, spec.with_output, out_space);
    if spec.indirect_source {
        op.0.write().unwrap().flags |= pcodeop_flags::INDIRECT_SOURCE;
    }
    let input = op.0.read().unwrap().get_in(0).unwrap().clone();
    let output = op.0.read().unwrap().output.clone();
    if let Some(output) = &output {
        let mut value = output.write().unwrap();
        if spec.write_mask {
            value.set_write_mask();
        }
        if spec.auto_live {
            value.set_auto_live_hold();
        }
    }
    if spec.descendant {
        let output = output.as_ref().expect("descendant case requires output");
        let sink = graph.new_op(OpCode::CPUI_COPY, 1, true, AddressSpace::Unique);
        graph.fd.op_set_input(&sink, output.clone(), 0);
    }

    let desc_before = output
        .as_ref()
        .map_or(-1, |value| value.read().unwrap().count_descends() as i32);
    let input_desc_before = input.read().unwrap().count_descends();
    let (call_before, indirect_before) = {
        let value = op.0.read().unwrap();
        (value.is_call(), value.is_indirect_source())
    };
    let write_before = output
        .as_ref()
        .is_some_and(|value| value.read().unwrap().is_write_mask());
    let auto_before = output
        .as_ref()
        .is_some_and(|value| value.read().unwrap().is_auto_live());
    let space_deadcode = output
        .as_ref()
        .is_some_and(|value| value.read().unwrap().get_space().does_deadcode());

    let result = RuleEarlyRemoval::new()
        .apply_op(&op.0, &mut graph.fd)
        .expect("RuleEarlyRemoval apply");
    let input_desc_after = input.read().unwrap().count_descends();
    let alive = bank_contains_op(&graph.fd.obank.alivelist, &op.0);
    let dead = bank_contains_op(&graph.fd.obank.deadlist, &op.0);
    let block_member = block_contains(&graph.block, &op.0);
    let output_present = output
        .as_ref()
        .is_some_and(|value| bank_contains_vn(&graph.fd, value));
    let output_attached = op.0.read().unwrap().output.is_some();

    println!(
        "case={}|result={}|pass={}|delay={}|call={}|indirect={}|output={}|desc={}|writemask={}|autolive={}|space_deadcode={}|deadremoved={}|alive={}|dead={}|block={}|output_present={}|output_attached={}|input_desc={}->{}",
        spec.name,
        result,
        spec.pass,
        spec.dead_delay,
        usize::from(call_before),
        usize::from(indirect_before),
        usize::from(output.is_some()),
        desc_before,
        usize::from(write_before),
        usize::from(auto_before),
        usize::from(space_deadcode),
        dead_removed(&graph.fd, out_space),
        usize::from(alive),
        usize::from(dead),
        usize::from(block_member),
        usize::from(output_present),
        usize::from(output_attached),
        input_desc_before,
        input_desc_after,
    );

    if spec.name == "neither_allowed" {
        println!(
            "residual=nullable_input_arity|post_num_inputs={}|slot0_null=unrepresentable|status=MISMATCH",
            op.0.read().unwrap().num_input(),
        );
    }
}

fn opcode_list(opcodes: &[OpCode]) -> String {
    opcodes
        .iter()
        .map(|opcode| (*opcode as i32).to_string())
        .collect::<Vec<_>>()
        .join(",")
}

fn run_dispatch() {
    let mut graph = Graph::new("typed_dispatch", 0x7200, 1);
    graph
        .fd
        .heritage
        .set_dead_code_delay(AddressSpace::Unique, 0);
    let typed = RuleEarlyRemoval::new().get_opcodes();
    println!("typed_oplist={}|count={}", opcode_list(&typed), typed.len());
    println!(
        "residual=raw_dispatch|raw=unavailable|count=unavailable|untyped=0,45|status=MISMATCH"
    );

    for opcode in &typed {
        graph.new_op(*opcode, 0, true, AddressSpace::Unique);
    }
    let mut pool = ActionPool::with_flags("earlyremoval_dispatch", 0);
    pool.add_rule(Box::new(RuleEarlyRemoval::new()));
    let mut state = ActionState::new(0);
    let result = pool
        .perform(&mut graph.fd, &mut state)
        .expect("ActionPool perform");
    let rule_state = pool.rule_state(0).expect("registered rule state");
    let alive = graph
        .fd
        .obank
        .alivelist
        .iter()
        .map(|op| (op.0.read().unwrap().opcode as i32).to_string())
        .collect::<Vec<_>>()
        .join(",");
    println!(
        "dispatch|perform={}|status={}|tests={}|apply={}|alive={}|dead_count={}|deadremoved={}",
        result,
        state.status,
        rule_state.count_tests,
        rule_state.count_apply,
        alive,
        graph.fd.obank.deadlist.len(),
        dead_removed(&graph.fd, AddressSpace::Unique),
    );
}

fn main() {
    let cases = [
        CaseSpec {
            name: "call_guard",
            opcode: OpCode::CPUI_CALL,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "indirect_guard",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: true,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "no_output_guard",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: false,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "descendant_guard",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: true,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "autolive_only",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: true,
            other_space: false,
        },
        CaseSpec {
            name: "write_and_autolive",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: true,
            auto_live: true,
            other_space: false,
        },
        CaseSpec {
            name: "write_mask_only",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: true,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "neither_allowed",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "delay_zero_equal",
            opcode: OpCode::CPUI_COPY,
            pass: 0,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "delay_one_equal",
            opcode: OpCode::CPUI_COPY,
            pass: 1,
            dead_delay: 1,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "delay_one_after",
            opcode: OpCode::CPUI_COPY,
            pass: 2,
            dead_delay: 1,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: false,
        },
        CaseSpec {
            name: "no_deadcode_other",
            opcode: OpCode::CPUI_COPY,
            pass: 0,
            dead_delay: 0,
            with_output: true,
            indirect_source: false,
            descendant: false,
            write_mask: false,
            auto_live: false,
            other_space: true,
        },
    ];
    for (index, case) in cases.into_iter().enumerate() {
        run_case(case, 0x7000 + index as u64 * 0x20);
    }
    run_dispatch();
    println!(
        "residual=heritage_manager_projection|scope=full_manager_state_and_late_generation_warning|status=UNTESTED"
    );
}
